Optimize storage of previous transactions

j-matus opened the pull request.

j-matus committed:
cardano: use less memory

j-matus committed:
cardano: do not create unnecessary lists

refi93 commented on src/apps/cardano/sign_tx.py:
```diff
@@ -74,12 +74,16 @@ def is_change(output, inputs):
             tx_ack = await request_transaction(ctx, tx_req, index)
-            transactions.append(tx_ack.transaction)
+            tx_hash = hashlib.blake2b(data=bytes(tx_ack.transaction), outlen=32).digest()
+            transaction.tx_data[tx_hash] = cbor.decode(tx_ack.transaction)
```
hmm, if I understand right, this ends up storing all decoded input transactions in memory at once. We could avoid that, the idea is to set a flag "attested" for each of the msg.inputs, that have a hash, index and amount matching with the currently decoded transaction. That way you can forget that decoded transaction right after processing it. This should yield significantly better results than the current approach

j-matus committed:
cardano: refactor input processing

Refactor input processing. Instead of creating a dict, compare the hash
of every transaction to the hash of every input since space complexity
is crucial.

refi93 commented on src/apps/cardano/sign_tx.py:
```diff
@@ -214,32 +192,30 @@ def _build_witnesses(self, tx_aux_hash: str):
             witnesses.append(
                 [
-                    self.types[index],
+                    (input.type or 0),
```
why `input.type or 0`? same below

j-matus replied:
See https://github.com/trezor/trezor-core/blob/master/src/apps/cardano/sign_tx.py , I just rewrote the code without the `types` list.

refi93 replied:
hmm interesting. I would rather drop the `or 0` then. I looked at the protobuf definition and it defaults to `0` anyway

refi93 replied:
ok, perhaps let's keep it there for the sake of readability: https://github.com/trezor/connect/blob/c7fb0032a8475cc09ec29ccc860fbe76066f4169/src/js/types/trezor.js#L601 in js the type field seems is optional so one might wonder to what it does default and this line of code would make that clear with the `or 0`. I've found that pattern also in other parts of the codebase

refi93 reviewed (changes requested):
see comments

refi93 commented on src/apps/cardano/sign_tx.py:
```diff
@@ -73,13 +73,30 @@ def is_change(output, inputs):
+                   amount = outputs[input.prev_index][1]
+                   input_coins += amount
+
+        for index in range(len(msg.inputs)):
```
This could be simplified to
```
if not all(attested):
   raise wire.ProcessError("No tx data sent for input " + index.find(False))
```

refi93 replied:
This could be simplified like this:
```
if (not all(attested)):
    raise wire.ProcessError("No tx data sent for input " + attested.index(False))
```

refi93 commented on src/apps/cardano/sign_tx.py:
```diff
@@ -73,13 +73,30 @@ def is_change(output, inputs):
+                   attested[i] = True
+                   outputs = tx_decoded[1]
+                   amount = outputs[input.prev_index][1]
+                   input_coins += amount
```
if you send the same transaction multiple times, input_coins will increase multiple times. I would suggest rewriting the condition to
```
if (not attested[i]) and (bytes(input.prev_hash) == tx_hash):
```

refi93 commented on src/apps/cardano/sign_tx.py:
```diff
@@ -73,13 +73,30 @@ def is_change(output, inputs):
-            transactions.append(tx_ack.transaction)
+            tx_hash = hashlib.blake2b(data=bytes(tx_ack.transaction), outlen=32).digest()
+            tx_decoded = cbor.decode(tx_ack.transaction)
+            for i, input in enumerate(msg.inputs):
```
if you send the same transaction multiple times, input_coins will increase multiple times. I would suggest rewriting the condition to
```
if (not attested[i]) and (bytes(input.prev_hash) == tx_hash):
```

refi93 commented on src/apps/cardano/sign_tx.py:
```diff
@@ -73,13 +73,30 @@ def is_change(output, inputs):
+            tx_hash = hashlib.blake2b(data=bytes(tx_ack.transaction), outlen=32).digest()
+            tx_decoded = cbor.decode(tx_ack.transaction)
+            for i, input in enumerate(msg.inputs):
+                if bytes(input.prev_hash) == tx_hash:
```
if you send to Trezor the same transaction multiple times, input_coins will increase multiple times. I would suggest rewriting the condition to
```
if (not attested[i]) and (bytes(input.prev_hash) == tx_hash):
```

refi93 commented on src/apps/cardano/sign_tx.py:
```diff
@@ -119,56 +133,19 @@ def __init__(
-        transactions: list,
         keychain,
         protocol_magic: int,
+        input_coins: int,
```
I would rename that to input_coins_sum to make clear that it is a single number

j-matus committed:
cardano: changes after review
